rust/src/rfb/rfb.rs

```diff
@@ -0,0 +1,851 @@
+        if self.events != std::ptr::null_mut() {
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for RFBTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct RFBState {
+    tx_id: u64,
+    request_buffer: Vec<u8>,
+    response_buffer: Vec<u8>,
+    transactions: Vec<RFBTransaction>,
+    state: parser::RFBGlobalState
+}
+
+impl RFBState {
+    pub fn new() -> Self {
+        Self {
+            tx_id: 0,
+            request_buffer: Vec::new(),
+            response_buffer: Vec::new(),
+            transactions: Vec::new(),
+            state: parser::RFBGlobalState::TCServerProtocolVersion
+        }
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&RFBTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self) -> RFBTransaction {
+        let mut tx = RFBTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn get_current_tx(&mut self) -> Option<&mut RFBTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == self.tx_id {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> bool {
+        // We're not interested in empty requests.
+        if input.len() == 0 {
+            return true;
+        }
+
+        // For simplicity, always extend the buffer and work on it.
+        self.request_buffer.extend(input);
+
+        let tmp: Vec<u8>;
+        let mut current = {
+            tmp = self.request_buffer.split_off(0);
+            tmp.as_slice()
+        };
+
+        SCLogDebug!("request_state {}, input_len {}", self.state, input.len());
+        loop {
+            if current.len() == 0 {
+                return true;
+            }
+            match self.state {
+                parser::RFBGlobalState::TSClientProtocolVersion => {
+                    match parser::parse_protocol_version(current) {
+                        Ok((rem, request)) => {
+                            current = rem;
+                            if request.major.to_string() == "003" && request.minor.to_string() == "003" {
```

Comment:
Should be able to compare without the .to_string() here (avoids short lived copies).